octodns_constellix/__init__.py

```diff
@@ -178,59 +183,115 @@ def record_delete(self, zone_name, record_type, record_id):
         path = f'/domains/{zone_id}/records/{record_type}/{record_id}'
         return self._request('DELETE', path).json()
 
-    def pools(self, pool_type):
-        if self._pools[pool_type] is None:
+    def pool_update_cache(self, pool_type, pool_id, value):
+        # Create cache for pool_type if needed
+        if value is not None and self._pools[pool_type] is None:
             self._pools[pool_type] = {}
+
+        if self._pools[pool_type] is not None:
+            self._pools[pool_type][pool_id] = value
+
+    def pool_clear_cache(self, pool_type, pool_id):
+        self.log.debug('pool_clear_cache %s %d', pool_type, pool_id)
+        self.pool_update_cache(pool_type, pool_id, None)
+
+    def pool_get_cache(self, pool_type):
+        if self._pools[pool_type] is None:
+            return []
+        else:
+            return self._pools[pool_type].values()
+
+    def pools(self, pool_type):
+        self.log.debug('pools %s', pool_type)
+        # The list view does not return checkId, policy and other details
+        result = self.pool_get_cache(pool_type)
+        if 0 == len(self.pool_get_cache(pool_type)):
             path = f'/pools/{pool_type}'
             response = self._request('GET', path).json()
             for pool in response:
-                self._pools[pool_type][pool['id']] = pool
-        return self._pools[pool_type].values()
+                self.pool_update_cache(pool_type, pool['id'], pool)
+            result = self.pool_get_cache(pool_type)
+        return result
 
-    def pool(self, pool_type, pool_name):
+    def pool_by_id(self, pool_type, pool_id):
+        self.log.debug('pool_by_id %s %d', pool_type, pool_id)
         pools = self.pools(pool_type)
+        result = None
         for pool in pools:
-            if pool['name'] == pool_name and pool['type'] == pool_type:
-                return pool
-        return None
+            if pool['id'] == pool_id:
+                result = pool
+                break
 
-    def pool_by_id(self, pool_type, pool_id):
+        # We do not have the details yet, fetch the pool.
+        # Do not try to fetch details for something that is not in the list.
+        if result is not None and result.get('failedFlag', None) is None:
+            path = f'/pools/{pool_type}/{pool_id}'
+            result = self._request('GET', path).json()
+            self._pools[pool_type][pool['id']] = result
+
+        return result
+
+    def pool_by_name(self, pool_type, pool_name):
+        self.log.debug('pool_by_name %s %s', pool_type, pool_name)
         pools = self.pools(pool_type)
         for pool in pools:
-            if pool['id'] == pool_id:
-                return pool
+            if pool['name'] == pool_name and pool['type'] == pool_type:
```

Comment:
Looks like the 2nd condition will always be true because we're already iterating on pools of that given type only?